### PR TITLE
Fix handling of trailing backslash in regular expressions

### LINKED ISSUE #117853: A regexp with a trailing backslash (invalid in re2) is silently answered as a substring search by `match`, `extractAll`, `countMatches`, and `splitByRegexp` — a lone `'\\'` pattern matches every row, while the same pattern throws in `replaceRegexpAll`, `multiMatchAny`, and any non-trivial spelling
### Company or project name

ClickHouse (durability testing)

### Describe what's wrong

A regular expression that ends in an unescaped trailing backslash is invalid in re2 (`trailing \` is a compile error). But when such a pattern is otherwise "trivial" (no metacharacters), ClickHouse never consults re2 at all: the pattern analyzer `OptimizedRegularExpression::analyzeImpl` consumes the dangling backslash at the very end of the pattern and simply stops (`src/Common/OptimizedRegularExpression.cpp:349-351` — `++pos; if (pos == end) break;`), leaving the pattern classified as a trivial literal. The query is then answered with a plain substring search, silently dropping the trailing backslash from the pattern.

The result is inconsistent three ways on the exact same invalid pattern (all examples below use the regexp `abc\`, written `'abc\\'` as an SQL literal):

1. `match`, `extractAll`, `countMatches`, and `splitByRegexp` accept it and answer it as if the pattern were `abc`. The degenerate case is a pattern consisting of a lone backslash: `match(s, '\\')` reduces to the empty substring and returns 1 for every row.
2. `replaceRegexpAll` / `replaceRegexpOne` throw `BAD_ARGUMENTS` ("The pattern argument is not a valid re2 pattern: trailing \\"), and `multiMatchAny` throws `BAD_ARGUMENTS` ("Unescaped \\ at end of input") — the same pattern, rejected loudly by the sibling functions.
3. Any non-trivial spelling of the same pattern makes `match` itself throw: `match(s, 'ab(c)\\')` and `match(s, '^abc\\')` both raise `CANNOT_COMPILE_REGEXP`, because those patterns reach re2.

Besides masking the error, this returns wrong rows for the most likely user intent: someone who wants to match a literal backslash after `abc` and under-escapes (`abc\` instead of `abc\\`) gets back every row containing plain `abc` — with no error to tell them the pattern was invalid. `LIKE`/`ILIKE` already handle the analogous case correctly (`'abcd' LIKE 'abc\\'` throws `CANNOT_PARSE_ESCAPE_SEQUENCE`).

### Does it reproduce on the most recent release?

Yes. Verified on `26.4.1.1` and on a master build `26.9.1.1`; the responsible code in `src/Common/OptimizedRegularExpression.cpp` is unchanged on current master (`dbed019389ee`), including after #117331 (which extended escape handling but only for escapes with a character after the backslash).

### How to reproduce

Default settings, any table or constant. Note the SQL string escaping: the SQL literal `'abc\\'` denotes the four-character regexp `abc\`.

```sql
SELECT match('abcd', 'abc\\');            -- 1        (expected: exception; the regexp is invalid)
SELECT match('abcd', '\\');               -- 1        (a lone backslash matches every string)
SELECT extractAll('abcd', 'abc\\');       -- ['abc']
SELECT countMatches('abcabc', 'abc\\');   -- 2
SELECT splitByRegexp('abc\\', 'xabcy');   -- ['x','y']

-- The same invalid pattern is rejected everywhere else:
SELECT match('abcd', 'ab(c)\\');                 -- CANNOT_COMPILE_REGEXP: trailing \
SELECT match('abcd', '^abc\\');                  -- CANNOT_COMPILE_REGEXP: trailing \
SELECT replaceRegexpAll('abcd', 'abc\\', 'X');   -- BAD_ARGUMENTS: trailing \
SELECT multiMatchAny('abcd', ['abc\\']);         -- BAD_ARGUMENTS: Unescaped \ at end of input
```

### Expected behavior

The functions consistently reject the invalid regexp with `CANNOT_COMPILE_REGEXP`, the same way `match` already does for any non-trivial pattern with the same trailing backslash.

### Error message and/or stacktrace

No error is produced on the affected paths — that is the bug. The sibling paths produce:

```
Code: 427. DB::Exception: OptimizedRegularExpression: cannot compile re2: ^abc\, error: trailing \. (CANNOT_COMPILE_REGEXP)
Code: 36. DB::Exception: The pattern argument is not a valid re2 pattern: trailing \. (BAD_ARGUMENTS)
```

### Additional context

The analyzer's escape handler advances past the backslash and, if that was the last byte of the pattern, breaks out of the token loop without marking the pattern non-trivial and without recording an error. Everything the analyzer classified as trivial is answered by substring search in `src/Functions/MatchImpl.h`, so re2 (which would reject the pattern) never sees it. Escapes with a following character are handled correctly (identity escapes since #117331, and `\x`-style escapes are consumed whole since #106382 was fixed) — only the dangling terminal backslash falls through.

### PR DESCRIPTION
**Pull request by alexprabhat99**
Closes: https://github.com/ClickHouse/ClickHouse/issues/117853

### Changelog category (leave one):

* Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a [[user-readable short description](https://github.com/ClickHouse/ClickHouse/blob/master/docs/changelog_entry_guidelines.md)] of the changes that goes into CHANGELOG.md):

Fix handling of regular expressions ending with an unescaped backslash in match, extractAll, countMatches, and splitByRegexp. Such invalid patterns are no longer incorrectly treated as trivial substring matches.

---
Workflow [[PR](https://s3.amazonaws.com/clickhouse-test-reports/praktika.html?PR=118229&sha=latest&name_0=PR)]
Sync PR [[sync-upstream/pr/118229](https://github.com/search?q=head%3Async-upstream%2Fpr%2F118229+org%3AClickHouse+type%3Apr&type=pullrequests)]

### CODE REVIEW

File: src/Storages/MergeTree/MergeTreeIndexBloomFilterText.cpp [hunk 1]
```diff
@@ -790,6 +791,9 @@ bool MergeTreeConditionBloomFilterText::traverseTreeEquals(
         out.bloom_filter = std::make_unique<BloomFilter>(params);
 
         auto & value = const_value.safeGet<String>();
+        /// Compile the pattern as `match` execution does, so an invalid regexp
+        /// raises an exception instead of being silently pruned.
```
Comment (rschu1ze) [inline src/Storages/MergeTree/MergeTreeIndexBloomFilterText.cpp]: Future readers will not understand what "instead of being silently pruned" means. Could you rewrite (clarify), please?

Comment (rschu1ze) [reply src/Storages/MergeTree/MergeTreeIndexBloomFilterText.cpp]: Please check if these other skip index types are also affected:

[bloom_filter](https://clickhouse.com/docs/reference/engines/table-engines/mergetree-family/mergetree#bloom-filter) index
[ngrambf_v1](https://clickhouse.com/docs/reference/engines/table-engines/mergetree-family/mergetree#n-gram-bloom-filter) index
[tokenbf_v1](https://clickhouse.com/docs/reference/engines/table-engines/mergetree-family/mergetree#token-bloom-filter) index
[text](https://clickhouse.com/docs/reference/engines/table-engines/mergetree-family/mergetree#text) index

File: src/Common/OptimizedRegularExpression.cpp [hunk 2]
```diff
@@ -348,7 +348,10 @@ const char * analyzeImpl(
             {
                 ++pos;
                 if (pos == end)
+                {
+                    is_trivial = false;
```
Comment (rschu1ze) [inline src/Common/OptimizedRegularExpression.cpp]: Is this not the place where we should throw an exception?

**Comment by alexprabhat99** [comment]
Hi @PedroTadim 
All CI checks have passed successfully. Could you please review this when you have a chance? Thank you.

File: tests/queries/0_stateless/01866_split_by_regexp.sql [hunk 3]
```diff
@@ -19,7 +19,7 @@ select splitByRegexp(']', 'a]b]c');
 select splitByRegexp('{', 'a{b{c');
 select splitByRegexp('}', 'a}b}c');
 select splitByRegexp('|', 'a|b|c');
-select splitByRegexp('\\', 'a\\b\\c');
+select splitByRegexp('\\', 'a\\b\\c'); -- { serverError CANNOT_COMPILE_REGEXP }
```
Comment (PedroTadim) [inline tests/queries/0_stateless/01866_split_by_regexp.sql:22]: 🕵️ One thing to settle first. The changelog says patterns are rejected with `CANNOT_COMPILE_REGEXP`, and all the assertions expect that, but on master the same pattern through `replaceRegexpAll` gives `BAD_ARGUMENTS` (code 36, "trailing \"). So after this we have consistent *rejection* but two different error codes for the same invalid regexp depending on the function.

File: tests/queries/0_stateless/05059_regexp_trailing_backslash.sql [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+-- A trailing unescaped backslash is an invalid RE2 regexp and must not
+-- be treated as a trivial substring.
+SELECT match('abcd', 'abc\\'); -- { serverError CANNOT_COMPILE_REGEXP }
+SELECT match('abcd', '\\'); -- { serverError CANNOT_COMPILE_REGEXP }
+SELECT extractAll('abcd', 'abc\\'); -- { serverError CANNOT_COMPILE_REGEXP }
+SELECT countMatches('abcabc', 'abc\\'); -- { serverError CANNOT_COMPILE_REGEXP }
+SELECT splitByRegexp('abc\\', 'xabcy'); -- { serverError CANNOT_COMPILE_REGEXP }
+
+-- Valid escaped backslashes must continue working.
+SELECT match('abc\\', 'abc\\\\');
```
Comment (PedroTadim) [inline tests/queries/0_stateless/05059_regexp_trailing_backslash.sql:21]: @alexprabhat99 can you add all test cases from the issue here?

File: src/Storages/MergeTree/MergeTreeIndexBloomFilterText.cpp [hunk 5]
```diff
@@ -790,6 +791,9 @@ bool MergeTreeConditionBloomFilterText::traverseTreeEquals(
...
+        Regexps::createRegexp</*like=*/ false, /*no_capture=*/ true, /*case_insensitive=*/ false>(value);
```
Comment (PedroTadim) [inline src/Storages/MergeTree/MergeTreeIndexBloomFilterText.cpp:853]: 🕵️ 

On the validation call in `traverseTreeEquals`:

```cpp
Regexps::createRegexp</*like=*/ false, /*no_capture=*/ true, /*case_insensitive=*/ false>(value);
RegexpAnalysisResult result = OptimizedRegularExpression::analyze(value);
```

The behaviour is right, but this compiles a regexp, discards it, then processes the same string again. `Regexps.h` has `LocalCacheTable::getOrSet` for exactly this, with a comment noting it exists to avoid recompiling the small set of patterns a query evaluates repeatedly — calling the raw `createRegexp` bypasses it, so no cache hit is possible and neither `RegexpLocalCacheHit` nor `RegexpLocalCacheMiss` fires, which also makes the cost invisible in profiling. Using the cache and binding the result would get the throw-on-invalid behaviour for free:

```cpp
static thread_local Regexps::LocalCacheTable cache;
auto regexp = cache.getOrSet</*like=*/ false, /*no_capture=*/ true, /*case_insensitive=*/ false>(value);
```

This is index analysis, so it runs once per `match` predicate per index rather than per row — the absolute cost is small, and I am not blocking on it. It just seems like avoidable waste next to a purpose-built cache.

One thing worth checking rather than assuming: the comment says it compiles "as `match` execution does", but this passes `RE_DOT_NL | RE_NO_CAPTURE` with `case_insensitive = false`. If execution can reach here with different flags (an `iLike`-derived pattern, say), a pattern could validate under one flag set and behave differently under another. Is the flag triple guaranteed to match what the consumer uses?

**Review by rschu1ze — CHANGES_REQUESTED** [review]

File: src/Common/OptimizedRegularExpression.cpp (hunk 2 above)
Comment (alexprabhat99) [reply src/Common/OptimizedRegularExpression.cpp]: I checked this path. `OptimizedRegularExpression::analyze()` catches exceptions from `analyzeImpl()` and returns an empty analysis result, so throwing directly here would not propagate the regexp error. I plan to keep this as `is_trivial = false`, which makes the normal execution path reach regexp compilation and reject the invalid pattern.

File: src/Storages/MergeTree/MergeTreeIndexBloomFilterText.cpp (hunk 1 above)
Comment (alexprabhat99) [reply src/Storages/MergeTree/MergeTreeIndexBloomFilterText.cpp]: > Future readers will not understand what "instead of being silently pruned" means. Could you rewrite (clarify), please?

Agreed. I’ll reword this to clarify that the regexp is validated before its required substring is used to build the skip-index condition.

Comment (alexprabhat99) [reply src/Storages/MergeTree/MergeTreeIndexBloomFilterText.cpp]: > Please check if these other skip index types are also affected:
> 
> [bloom_filter](https://clickhouse.com/docs/reference/engines/table-engines/mergetree-family/mergetree#bloom-filter) index [ngrambf_v1](https://clickhouse.com/docs/reference/engines/table-engines/mergetree-family/mergetree#n-gram-bloom-filter) index [tokenbf_v1](https://clickhouse.com/docs/reference/engines/table-engines/mergetree-family/mergetree#token-bloom-filter) index [text](https://clickhouse.com/docs/reference/engines/table-engines/mergetree-family/mergetree#text) index

Checked all four. `ngrambf_v1` and `tokenbf_v1` share `MergeTreeConditionBloomFilterText`, so both are affected and covered by the same fix; I’ll add regression coverage for `tokenbf_v1` as well. The generic `bloom_filter` condition does not use this `match` regexp-analysis path. The `text` index has a separate `match` path, but it already explicitly compiles the regexp before deriving index tokens, so the invalid pattern is already rejected there.

File: tests/queries/0_stateless/05059_regexp_trailing_backslash.sql (hunk 4 above)
Comment (alexprabhat99) [reply tests/queries/0_stateless/05059_regexp_trailing_backslash.sql:21]: Yes. I’ll add all reproducer cases from #117853, including the non-trivial `match` patterns and the existing `replaceRegexpAll` / `multiMatchAny` rejection cases.

File: src/Storages/MergeTree/MergeTreeIndexBloomFilterText.cpp (hunk 5 above)
Comment (alexprabhat99) [reply src/Storages/MergeTree/MergeTreeIndexBloomFilterText.cpp:853]: Agreed. I’ll switch this validation to `Regexps::LocalCacheTable::getOrSet()` so repeated validation can use the existing regexp cache and its profiling counters.

Comment (alexprabhat99) [reply src/Storages/MergeTree/MergeTreeIndexBloomFilterText.cpp:853]: Checked the call path. This validation is only inside the `function_name == "match"` branch; `iLike` does not reach it. Therefore `like=false`, `no_capture=true`, and `case_insensitive=false` match the consumer here.

File: tests/queries/0_stateless/01866_split_by_regexp.sql (hunk 3 above)
Comment (alexprabhat99) [reply tests/queries/0_stateless/01866_split_by_regexp.sql:22]: Good point. `replaceRegexpAll` already reports `BAD_ARGUMENTS` for this malformed pattern on master, while the affected functions in #117853 report `CANNOT_COMPILE_REGEXP` once they reach regexp compilation. I’ll preserve the existing per-function error codes rather than broaden this fix into error-code standardization, add the `replaceRegexpAll` case to the regression coverage, and adjust the changelog to state that these invalid patterns are rejected rather than guaranteeing a specific error code across all regexp functions.

File: tests/queries/0_stateless/05059_regexp_trailing_backslash.sql [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    s String,
+    INDEX idx s TYPE tokenbf_v1(512, 2, 0)
+)
+ENGINE = MergeTree
+ORDER BY tuple()
+SETTINGS index_granularity = 1;
+
+INSERT INTO regexp_token_idx VALUES ('nothing here');
+
+SELECT count()
```
Comment (rschu1ze) [inline tests/queries/0_stateless/05059_regexp_trailing_backslash.sql:61]: @alexprabhat99 Could you please add a test for [the full-text index](https://clickhouse.com/docs/reference/engines/table-engines/mergetree-family/textindexes)
1. with a splitByRegexp tokenizer with an invalid regexp. This should fail at index creation time.
2. with the default tokenizer (splitByNonAlpha) and a SELECT query with a `WHERE match(s, 'abc\\')` clause

?

File: src/Storages/MergeTree/MergeTreeIndexBloomFilterText.cpp [hunk 7]
```diff
@@ -847,6 +848,11 @@ bool MergeTreeConditionBloomFilterText::traverseTreeEquals(
         out.bloom_filter = std::make_unique<BloomFilter>(params);
 
         auto & value = const_value.safeGet<String>();
+        /// Validate the regexp before using its required substring to build
+        /// the skip-index condition.
+        static thread_local Regexps::LocalCacheTable regexp_cache;
```
Comment (rschu1ze) [inline src/Storages/MergeTree/MergeTreeIndexBloomFilterText.cpp]: I don't think we need a local cache table, at least not in the scope of this PR. If there is a performance problem it fixes, then we can implement that in a separate PR.

File: src/Common/OptimizedRegularExpression.cpp (hunk 2 above)
Comment (rschu1ze) [inline src/Common/OptimizedRegularExpression.cpp]: https://github.com/ClickHouse/ClickHouse/pull/118634/ implements this place slightly differently. Which version is right?

Comment (alexprabhat99) [reply src/Common/OptimizedRegularExpression.cpp]: Thanks. I've aligned this with #118634 and now use `finish_non_trivial_char()` for a trailing lone backslash. This keeps the pattern on the non-trivial regexp path and lets RE2 reject it normally.

File: src/Storages/MergeTree/MergeTreeIndexBloomFilterText.cpp (hunk 7 above)
Comment (alexprabhat99) [reply src/Storages/MergeTree/MergeTreeIndexBloomFilterText.cpp]: Agreed. Removed the local cache and kept the validation direct. Any caching optimization can be handled separately if needed.

File: tests/queries/0_stateless/05059_regexp_trailing_backslash.sql (hunk 6 above)
Comment (alexprabhat99) [reply tests/queries/0_stateless/05059_regexp_trailing_backslash.sql:61]: Added both requested full-text index tests: one for an invalid `splitByRegexp` tokenizer at index creation time, and one using the default `splitByNonAlpha` tokenizer with an invalid regexp in `match()`.

**Review by rschu1ze — APPROVED** [review]

File: tests/queries/0_stateless/05059_regexp_trailing_backslash.sql [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+    s String,
+    INDEX idx s TYPE text(tokenizer = splitByNonAlpha)
+)
+ENGINE = MergeTree
+ORDER BY tuple()
+SETTINGS index_granularity = 1;
+
+INSERT INTO regexp_text_idx VALUES ('nothing here');
+
+SELECT count()
```
Comment (PedroTadim) [inline tests/queries/0_stateless/05059_regexp_trailing_backslash.sql:95]: @alexprabhat99 the `multiMatchAny` function case seems to be missing. Can you add it to confirm the expected behavior?

Comment (PedroTadim) [reply tests/queries/0_stateless/05059_regexp_trailing_backslash.sql:95]: Ok, we can ignore this edge case.
